### ISSUE TITLE
Complete rewrite of gatherer needed (again)

### ISSUE THREAD

**Issue by ReidWeb** [XIVStats/XIVStats#12]
# Background

To give some background: in November/December of 2015, Pricetx noted to me the 'hastle' of having to run multiple instances of his ruby script to get a 'fast census'. I picked up on this, and I was intending to learn threading in Java anyway - so I took it as an opportunity to learn and implement something useful at the same time. Unfortunately at the time I was not greatly experienced in testing techniques, database connectivity or handling errors in the best manner. It has also become evident to me that Java may not have been the best choice too.

# Flaws in current implementation

- badly written, hard to maintain database code
- no proper handling of database write errors
- handling of non-200 or 404 status codes is non-optimal - it just repeats until it gets a hit
- the 'parser' is highly coupled to the 'interpreter' of the data
- testing is flaky and awfully done

# Targets for rewrite

- scalable, maintainable database code
- properly handle database errors
- handle non-200, non-404 status codes in a better manner
- decouple 'parser' from 'interpreter'
- better testing - incorporating mocking technologies to allow for removal of reliance on data that might change. 

# Technology choice for rewrite

At present I am currently considering the following:

- JavaScript (Node.js) - this is the language I'm most comfortable in - and I've already PoCed this previously
- C# (.NET Core) - i'm relatively comfortable and do have a decent level of experience in C#, and I'm interested in developing out my skills in this area. Indications of performance of .NET Core and C# are very promising.
- Java rewrite - complete rewrite of the existing implementation from the ground up using the above targets.
- Python - no experience what so ever in Python, but interested to evaluate.

# Technology requirements

- allows for 'modular' code
- performance must be better than current Java implementation
- well maintained, singular HTTP library that supports `GET` and `HEAD` HTTP operations
- MySQL database support
- MongoDB database support - *need to further evaluate MongoDB performance for our use case, and consider alternatives in NoSQL space*
- runs on Windows, MacOS, Linux, FreeBSD - the runtime must be included in the package repos on Debian & CentOS and likewise for FreeBSD. On MacOS inclusion in brew or an alternative such as `nvm` would be preferable.
- well documented and maintained testing and mocking libraries to allow for improvement of testing

**Comment by Pricetx** [comment]
This issue pertains  to the Java Gatherer, but it's fine to keep it here.

It is worth noting that since this was originally raised, the Java application has been almost completely rewritten by @matthewhillier to a much better standard than previous. While there is no real reason why a rival implementation would be a bad thing, it's not necessarily as urgent as it once was. The frontend could probably do with re-designing though.

However, I assume the timeframe was for 2019 so I look forward to seeing this take shape in June.

**Comment by ReidWeb** [comment]
Yeah, would at least like to see a refactor that makes adding new "achievement recognitions" more decoupled from the code itself - would make @fahy's contributions much easier for him, I really appreciate his work to ensure we're updating them accordingly and it's definitely a nightmare to add them right now.

Front end could very much do with a rewrite, in React or similar.

**Comment by matthewhillier** [comment]
I'd be up for doing some work on the database structures. I've been playing around with adding new tables for gear and tracking over time, so it wouldn't be too much of a stretch to normalise out things like job levels, achievements, mounts, minions & gear - for example.

That would give us a lot more flexibility on what we can do on the front-end 👍

**Comment by Pricetx** [comment]
Tracking gear over time would actually be very interesting. It's probably the single most accurate way we could gauge an active player figure as this is the thing that would change the most often. 

But yeah, it really shouldn't be much work to support any changes to schema on the front-end, so i'm all in favour of any suggested improvements.

**Comment by ReidWeb** [comment]
Ref tracking gear over time, I'd a working version of this in the JS gatherer prototype i worked on. Poke me about this in June

**Comment by Almagnus1** [comment]
Any progress on this?

I'm looking at the active number, and they look really low compared to what Yoshi-P has said for sub numbers.

**Comment by fahy** [comment]
AFAIK, there hasn't been any progress. 

As of the sub numbers, Yoshida (Nor SE for that matter) have never revealed actual active subscribers. The only statistic that they reveal is how many accounts have been created.  The last figure that I have seen is about 16 million players.

For now, we are going to be updating the current site...We have made some adjustments towards checking stats for 5.0 but the true Shadowbringers figures won't be out will August (July is too early to get any meaningful stats)

We are looking to add more stats to the site and hoping that there is a minion or amount during the story during 5.0 to help us see what the active number of players are (The last 'active check' was Dress-up Raubahn from 4.1 story)
Some stuff will be delayed as we are going to be playing ShB :P
